File: BPNN Verification 2/bpnn-msv-2.m
frame(v,w,r,o,g,e,f,x,b,y,yc,Ek,aveEk) and (
float v[3,5] and skip;
float w[5,1] and skip;
float r[5] and skip;
float o[1] and skip;
float g[1] and skip;
float e[5] and skip;
float f[3] and skip;
float x[3] and skip;
float b[5] and skip;
float y[1] and skip;
float yc[1] and skip;
float Ek and skip;
float aveEk and skip;
struct _iobuf {
char *_ptr and 
int _cnt and 
char *_base and 
int _flag and 
int _file and 
int _charbuf and 
int _bufsiz and 
char *_tmpfname 
};
 function read_parameter ( int  RValue )
 {
     frame(read_parameter_in,read_parameter_buffer,read_parameter_i,read_parameter_h,read_parameter_h_1,read_parameter_j,read_parameter_h_3,read_parameter_j_1,return) and ( 
     int return<==0 and skip;
     FILE *read_parameter_in<==NULL and skip;
     read_parameter_in:=fopen("bpnn_param.txt","r");
     if(read_parameter_in=NULL) then 
     {
         fprintf(stderr,"Cannot open %s\n","bpnn_param.txt") and skip;
         return<==1 and RValue:=0;
         skip
         
     }
     else 
     {
          skip 
     };
     if(return=0)   then 
     {
         char read_parameter_buffer[256] and skip;
         int read_parameter_i<==0 and skip;
         
         while( (read_parameter_i<3) )
         {
             int read_parameter_h<==0 and skip;
             
             while( (read_parameter_h<5) )
             {
                 fscanf(read_parameter_in,"%s",read_parameter_buffer) and skip;
                 v[read_parameter_i,read_parameter_h]:=strtod(read_parameter_buffer,NULL,RValue);
                 read_parameter_h:=read_parameter_h+1
                 
             };
             read_parameter_i:=read_parameter_i+1
             
         };
         int read_parameter_h_1<==0 and skip;
         
         while( (read_parameter_h_1<5) )
         {
             int read_parameter_j<==0 and skip;
             
             while( (read_parameter_j<1) )
             {
                 fscanf(read_parameter_in,"%s",read_parameter_buffer) and skip;
                 w[read_parameter_h_1,read_parameter_j]:=strtod(read_parameter_buffer,NULL,RValue);
                 read_parameter_j:=read_parameter_j+1
                 
             };
             read_parameter_h_1:=read_parameter_h_1+1
             
         };
         int read_parameter_h_3<==0 and skip;
         
         while( (read_parameter_h_3<5) )
         {
             fscanf(read_parameter_in,"%s",read_parameter_buffer) and skip;
             r[read_parameter_h_3]:=strtod(read_parameter_buffer,NULL,RValue);
             read_parameter_h_3:=read_parameter_h_3+1
             
         };
         int read_parameter_j_1<==0 and skip;
         
         while( (read_parameter_j_1<1) )
         {
             fscanf(read_parameter_in,"%s",read_parameter_buffer) and skip;
             o[read_parameter_j_1]:=strtod(read_parameter_buffer,NULL,RValue);
             read_parameter_j_1:=read_parameter_j_1+1
             
         };
         fclose(read_parameter_in) and skip;
         return<==1 and RValue:=1;
         skip
     }
     else
     {
         skip
     }
     )
     }; 
  function bpnn_check_single (  )
 {
     frame(bpnn_check_single_ax,bpnn_check_single_by,bpnn_check_single_h,bpnn_check_single_alpha_h,bpnn_check_single_i,bpnn_check_single_j,bpnn_check_single_beta_j,bpnn_check_single_h_1,bpnn_check_single_j_1,bpnn_check_single_check_flag) and ( 
     float bpnn_check_single_ax[3]<=={0.991998,0.919734,0.859287} and skip;
     float bpnn_check_single_by[1]<=={0.923673} and skip;
	 // float bpnn_check_single_ax[3]<=={0.712532,0.451378,0.920685} and skip;
	 // float bpnn_check_single_by[1]<=={0.694865} and skip;
     int bpnn_check_single_h<==0 and skip;
     
     while( (bpnn_check_single_h<5) )
     {
         float bpnn_check_single_alpha_h<==0 and skip;
         int bpnn_check_single_i<==0 and skip;
         
         while( (bpnn_check_single_i<3) )
         {
             bpnn_check_single_alpha_h:=bpnn_check_single_alpha_h+v[bpnn_check_single_i,bpnn_check_single_h]*bpnn_check_single_ax[bpnn_check_single_i];
             bpnn_check_single_i:=bpnn_check_single_i+1
             
         };
         b[bpnn_check_single_h]:=(1/ (1+exp(-(bpnn_check_single_alpha_h-r[bpnn_check_single_h]))));
         bpnn_check_single_h:=bpnn_check_single_h+1
         
     };
     int bpnn_check_single_j<==0 and skip;
     
     while( (bpnn_check_single_j<1) )
     {
         float bpnn_check_single_beta_j<==0 and skip;
         int bpnn_check_single_h_1<==0 and skip;
         
         while( (bpnn_check_single_h_1<5) )
         {
             bpnn_check_single_beta_j:=bpnn_check_single_beta_j+w[bpnn_check_single_h_1,bpnn_check_single_j]*b[bpnn_check_single_h_1];
             bpnn_check_single_h_1:=bpnn_check_single_h_1+1
             
         };
         yc[bpnn_check_single_j]:=(1/ (1+exp(-(bpnn_check_single_beta_j-o[bpnn_check_single_j]))));
         bpnn_check_single_j:=bpnn_check_single_j+1
         
     };
     Ek:=0;
     int bpnn_check_single_j_1<==0 and skip;
     
     while( (bpnn_check_single_j_1<1) )
     {
         Ek:=Ek+(yc[bpnn_check_single_j_1]-bpnn_check_single_by[bpnn_check_single_j_1])*(yc[bpnn_check_single_j_1]-bpnn_check_single_by[bpnn_check_single_j_1]);
         bpnn_check_single_j_1:=bpnn_check_single_j_1+1
         
     };
     Ek:=0.5*Ek;
     int bpnn_check_single_check_flag<==0 and skip;
	 aveEk <==0.00000215 and skip;
     if(Ek<=aveEk) then 
     {
         bpnn_check_single_check_flag:=1
         
     }
     else 
     {
          skip 
     };
     if(bpnn_check_single_check_flag) then 
     {
         output ("Ek <= aveEk : ",Ek," <= ",aveEk," --> #success#","\n") and skip
     }
     else
     {
         output ("Ek > aveEk : ",Ek," > ",aveEk,"  -->  #failure#","\n") and skip
     }
     )
     }; 
  function main ( int  RValue )
 {
     frame(main_1_temp$_1,return) and (
     int return<==0 and skip;
     int main_1_temp$_1 and skip;
     main_1_temp$_1:=read_parameter(RValue);
     if(!main_1_temp$_1) then 
     {
         output ("Read param is failed in ","bpnn_param.txt","!\n") and skip;
         return<==1 and RValue:=0;
         skip
         
     }
     else 
     {
          skip 
     };
     if(return=0)   then 
     {
         bpnn_check_single();
         return<==1 and RValue:=0;
         skip
     }
     else
     {
         skip
     }
     )
 };
  main(RValue)
 )

File: BPNN Verification 2/bpnn-msv-2.c
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include <math.h>

#define IN_N                            3
#define OUT_N                           1
#define HIDDEN_N                        5
#define LEARN_RATE1                     0.3
#define LEARN_RATE2                     0.4

#define ACTIVATION_FUNC(x)              (1/(1+exp(-(x))))

#define SAVE_PARAM_PATH                 "../bpnn/dataset/bpnn_param.txt"

#define D   IN_N
#define Q   HIDDEN_N
#define L   OUT_N
#define U1  LEARN_RATE1
#define U2  LEARN_RATE2

#define SEEK_SET   0
#define RAND_MAX   32767

typedef int size_t;

typedef int bool;
#define true 1
#define false 0

double v[D][Q];
double w[Q][L];
double r[Q];
double o[L];

double g[L];
double e[Q];
double f[D]; // Modified by Yk

double x[D];
double b[Q];
double y[L];
double yc[L];

double Ek;
double aveEk = 0.00000215; // From help-pro-bak-2-msv

struct _iobuf {
	char *_ptr;
	int   _cnt;
	char *_base;
	int   _flag;
	int   _file;
	int   _charbuf;
	int   _bufsiz;
	char *_tmpfname;
	};
typedef struct _iobuf FILE;

/**
    读入神经网络结构参数
*/
bool read_parameter() {

    FILE* in = NULL;
    in = fopen(SAVE_PARAM_PATH, "r");

    if (in == NULL) {
        fprintf(stderr, "Cannot open %s\n", SAVE_PARAM_PATH);
        return false;
    }

    char buffer[256];

    for (size_t i = 0; i < D; i++) {
        for (size_t h = 0; h < Q; h++) {
            fscanf(in, "%s", buffer);
            v[i][h] = strtod(buffer, NULL);
        }
    }

    for (size_t h = 0; h < Q; h++) {
        for (size_t j = 0; j < L; j++) {
            fscanf(in, "%s", buffer);
            w[h][j] = strtod(buffer, NULL);
        }
    }

    for (size_t h = 0; h < Q; h++) {
        fscanf(in, "%s", buffer);
        r[h] = strtod(buffer, NULL);
    }

    for (size_t j = 0; j < L; j++) {
        fscanf(in, "%s", buffer);
        o[j] = strtod(buffer, NULL);
    }

    fclose(in);

    return true;
}

/**
    如何有效地检测异常样本？重新定义异常检测标准。
*/
void bpnn_check_single() {

    /* 定义输入层 */
    // failure
    double ax[3] = { 0.991998, 0.919734, 0.859287 };
    double by[1] = { 0.923673 };

    // success
    // double ax[3] = { 0.712532, 0.451378, 0.920685 };
    // double by[1] = { 0.694865 };

    /* 计算隐藏层的输出结果 */
    for (size_t h = 0; h < Q; h++) {
        double alpha_h = 0;
        for (size_t i = 0; i < D; i++)
            alpha_h += v[i][h] * ax[i];
        b[h] = ACTIVATION_FUNC(alpha_h - r[h]);
    }

    /* 计算输出层的输出结果 */
    for (size_t j = 0; j < L; j++) {
        double beta_j = 0;
        for (size_t h = 0; h < Q; h++)
            beta_j += w[h][j] * b[h];
        yc[j] = ACTIVATION_FUNC(beta_j - o[j]);
    }

    Ek = 0;

    /* 计算损失值*/
    for (size_t j = 0; j < L; j++) {
       Ek += (yc[j] - by[j]) * (yc[j] - by[j]);
    }
    Ek = 0.5 * Ek;

    // 利用样本的预测误差和实际生产环境下样本预测结果的平均误差进行对比
    // 如果前者小于后者，则认为样本为可靠样本，此处可视为“样本的正确性验证”。
    int check_flag = false; // 检测为失败
    if (Ek <= aveEk) {
        check_flag = true; // 检测为成功
    }

    if (check_flag) {
        printf("Ek <= aveEk : %0.8f <= %0.8f --> #success#", Ek, aveEk);
    } else {
        printf("Ek > aveEk : %0.8f > %0.8f  -->  #failure#", Ek, aveEk);
    }

    /* 样本输出结果的正确性验证

    define p0 : Ek != 0;
    define p1 : aveEk != 0;
    define q : Ek < aveEk;
    som (p0 and p1 and q)

    样本输出结果的正确性验证*/
}

int main()
{
    if (!read_parameter())
    {
        printf("Read param is failed in %s!\n", SAVE_PARAM_PATH);
        return 0;
    }

    // 第二步，在制定标准之后判断随机样本是否满足要求
    bpnn_check_single();

    return 0;
}
